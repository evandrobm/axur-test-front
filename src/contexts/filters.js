import React, { createContext, useState, useContext } from 'react'
import { node } from 'prop-types'

export const FiltersContext = createContext()

export const FiltersProvider = ({ children }) => {
  const [sortBy, setSortBy] = useState('desc')
  const [authorFilter, setAuthorFilter] = useState('')

  return (
    <FiltersContext.Provider
      value={{ sortBy, setSortBy, authorFilter, setAuthorFilter }}
    >
      {children}
    </FiltersContext.Provider>
  )
}

FiltersProvider.propTypes = {
  children: node
}

export const useFilters = () => useContext(FiltersContext)
