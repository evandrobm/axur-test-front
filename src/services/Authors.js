import axios from 'axios'

export const getAuthors = async () => {
  try {
    const { data } = await axios.get(
      'https://www.mocky.io/v2/5be5e3ae2f00005b000fc3f6'
    )
    return data
  } catch (err) {
    return false
  }
}
