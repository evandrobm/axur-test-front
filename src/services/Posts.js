import axios from 'axios'

export const getPosts = async () => {
  try {
    const { data } = await axios.get(
      'http://www.mocky.io/v2/5be5e3fa2f000082000fc3f8'
    )
    return data
  } catch (err) {
    return false
  }
}
