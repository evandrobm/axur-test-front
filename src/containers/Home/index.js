import React, { useState, useEffect, useMemo } from 'react'
import get from 'lodash/fp/get'

import { useFilters } from '../../contexts/filters'
import { getPosts } from '../../services/Posts'
import { getAuthors } from '../../services/Authors'

import PostListComponent from '../../components/PostList'
import FullPostsComponent from '../../components/FullPosts'

import { Container } from './styles'

const getDate = get('metadata.publishedAt')

const authorByPost = post => author =>
  get('id')(author) === get('metadata.authorId')(post)

const compareDates = sort => (a, b) =>
  sort === 'desc' ? getDate(b) - getDate(a) : getDate(a) - getDate(b)

const fetchPostsData = async (setPosts, setAuthors) => {
  const [postsRes, authorsRes] = await Promise.all([getPosts(), getAuthors()])
  setAuthors(authorsRes)
  setPosts(
    postsRes.sort(compareDates('desc')).map(post => ({
      ...post,
      author: authorsRes.find(authorByPost(post))
    }))
  )
}

const handleNavigation = post => () => {
  const reference = get('ref')(post)
  if (reference) {
    window.scrollTo(0, reference.offsetTop - 15)
  }
}

const Home = () => {
  const [posts, setPosts] = useState([])
  const [authors, setAuthors] = useState([])
  const { sortBy, authorFilter } = useFilters()

  const filteredPosts = useMemo(
    () =>
      posts
        .slice()
        .sort(compareDates(sortBy))
        .filter(post => {
          if (authorFilter === '') {
            return true
          }
          return get('author.id')(post) === parseInt(authorFilter)
        }),
    [posts, sortBy, authorFilter]
  )

  useEffect(() => {
    fetchPostsData(setPosts, setAuthors)
  }, [])

  return (
    <Container>
      <PostListComponent posts={posts} handleClick={handleNavigation} />
      <FullPostsComponent posts={filteredPosts} authors={authors} />
    </Container>
  )
}

export default Home
