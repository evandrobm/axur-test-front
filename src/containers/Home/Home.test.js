import React, { useState } from 'react'
import { render, screen, act } from '@testing-library/react'

import * as postsService from '../../services/Posts'
import * as authorsService from '../../services/Authors'
import { FiltersContext } from '../../contexts/filters'
import Home, { handleNavigation, compareDates } from './index'

const FiltersProvider = ({ children, defaultAuthor }) => {
  const [sortBy, setSortBy] = useState('desc')
  const [authorFilter, setAuthorFilter] = useState(defaultAuthor || '')

  return (
    <FiltersContext.Provider
      value={{ sortBy, setSortBy, authorFilter, setAuthorFilter }}
    >
      {children}
    </FiltersContext.Provider>
  )
}

const wrapper = ({ children }) => <FiltersProvider>{children}</FiltersProvider>
const wrapperFiltered = ({ children }) => (
  <FiltersProvider defaultAuthor="1">{children}</FiltersProvider>
)

const posts = [
  {
    title: 'Post Mock 1',
    body: 'Post body 1',
    metadata: { publishedAt: 1492004832000, authorId: 1 },
    ref: {
      offsetTop: 100
    }
  },
  {
    title: 'Post Mock 2',
    body: 'Post body 2',
    metadata: { publishedAt: 1492004832000, authorId: 2 }
  }
]

const authors = [
  {
    id: 1,
    name: 'Evandro'
  },
  {
    id: 2,
    name: 'Not Evandro'
  }
]

jest.mock('../../services/Posts')
jest.mock('../../services/Authors')

postsService.getPosts.mockResolvedValue(posts)
authorsService.getAuthors.mockResolvedValue(authors)

describe('Home container tests', () => {
  it('should handle navigation', () => {
    const scrollMock = jest.fn()
    window.scrollTo = scrollMock
    handleNavigation(posts[0])()
    expect(scrollMock).toHaveBeenCalledWith(0, 85)
    handleNavigation(posts[1])()
    expect(scrollMock).toHaveBeenCalledTimes(1)
  })
  it('should render posts', async () => {
    render(<Home />, { wrapper })
    const items = await screen.findAllByText(/Post Mock [.]*/)
    expect(items).toHaveLength(4)
  })
  it('should filter posts by author', async () => {
    render(<Home />, { wrapper: wrapperFiltered })
    const items = await screen.findAllByText(/Post Mock [.]*/)
    expect(items).toHaveLength(3)
  })
  it('should compare dates correctly', () => {
    const a = { metadata: { publishedAt: 1492004832000 } }
    const b = { metadata: { publishedAt: 1492004832001 } }
    const sortDesc = compareDates('desc')(a, b)
    const sortAsc = compareDates('asc')(a, b)
    expect(sortDesc > 0).toBeTruthy()
    expect(sortAsc < 0).toBeTruthy()
  })
  it('should match snapshot', async () => {
    await act(async () => {
      const { container } = render(<Home />, { wrapper })
      expect(container).toMatchSnapshot()
    })
  })
})
