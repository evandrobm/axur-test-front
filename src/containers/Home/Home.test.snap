// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container tests should match snapshot 1`] = `
<div>
  <main
    class="sc-fzoLag iiRdCn"
  >
    <aside
      class="sc-AxhCb jHmSzu"
    >
      <header
        class="sc-AxhUy grRGfi"
      >
        <h1
          class="sc-AxgMl dNFDPi"
        >
          Mocky Blog
        </h1>
      </header>
    </aside>
    <section
      class="sc-fzoyAV fwROSP"
    >
      <div
        class="sc-fznZeY fergZy"
      >
        <select
          class="sc-fzplWN sc-fznyAO NsLQw"
        >
          <option
            value=""
          >
            Select an author to filter...
          </option>
        </select>
        <select
          class="sc-fzplWN sc-fznKkj cxmOcm"
        >
          <option
            value="asc"
          >
            Newer
          </option>
          <option
            selected=""
            value="desc"
          >
            Older
          </option>
        </select>
      </div>
      <div
        class="sc-fzokOt gySKXb"
        width="750px"
      >
        <div
          class="sc-fzqNJr hQmfWt"
          height="35px"
          width="50%"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr kYizYw"
          height="14px"
          width="150px"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
      </div>
      <div
        class="sc-fzokOt gySKXb"
        width="750px"
      >
        <div
          class="sc-fzqNJr hQmfWt"
          height="35px"
          width="50%"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr kYizYw"
          height="14px"
          width="150px"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
      </div>
      <div
        class="sc-fzokOt gySKXb"
        width="750px"
      >
        <div
          class="sc-fzqNJr hQmfWt"
          height="35px"
          width="50%"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr kYizYw"
          height="14px"
          width="150px"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
        <div
          class="sc-fzqNJr fAiYJw"
        >
          <div
            class="sc-fzqBZW jYDWDu"
          />
        </div>
      </div>
    </section>
  </main>
</div>
`;
